### PR TITLE
[13.0][MIG]database_cleanup

### PR DESCRIPTION
**Pull request by AaronHForgeFlow**
Supersedes #1745 

Still need some cleanup (sorry for the joke), but it is usable.

Includes https://github.com/OCA/server-tools/pull/1822

### CODE REVIEW

File: database_cleanup/models/purge_models.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                except:
+                    pass
+            self.env["ir.model.relation"].search(
+                [("model", "=", line.name)]
+            ).with_context(**context_flags).unlink()
+            try:
+                self.env["ir.model"].browse([row[0]]).with_context(
+                    **context_flags
+                ).unlink()
+            except:
```
Comment (i-vyshnevska) [inline database_cleanup/models/purge_models.py]: this doesn't look like a good policy, I prefer previous except handling

File: database_cleanup/models/purge_modules.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        Uninstall modules upon manual confirmation, then reload
+        the database.
+        """
+        module_names = self.filtered(lambda x: not x.purged).mapped("name")
+        modules = self.env["ir.module.module"].search([("name", "in", module_names)])
+        if not modules:
+            return True
+        self.logger.info("Purging modules %s", ", ".join(module_names))
+        modules.filtered(
+            lambda x: x.state not in ("uninstallable", "uninstalled")
```
Comment (i-vyshnevska) [inline database_cleanup/models/purge_modules.py]: why removed " and x.name != "base"

**Review by i-vyshnevska — COMMENTED** [review]
black is still failing

**Comment by i-vyshnevska** [comment]
FTR some translation commits may be squashed

**Comment by AaronHForgeFlow** [comment]
@i-vyshnevska  thanks for reviewing. I agree with the changes proposed, will try to fix this soon.

**Comment by CasVissers-360ERP** [comment]
@AaronHForgeFlow Is #1396 also included in this PR?

**Comment by CasVissers-360ERP** [comment]
I also get a traceback when opening some menu items:
 ```
 File "/home/odoo/src/odoo/odoo/models.py", line 1747, in _add_missing_default_values
    defaults = self.default_get(list(missing_defaults))
  File "/home/odoo/src/user/community/database_cleanup/models/purge_wizard.py", line 45, in default_get
    res["purge_line_ids"] = self.find()
  File "/home/odoo/src/user/community/database_cleanup/models/purge_menus.py", line 50, in find
    menu.action.src_model and menu.action.src_model not in self.env
ValueError: <class 'AttributeError'>: "'ir.actions.act_window' object has no attribute 'src_model'" while evaluating
"action = env.get('cleanup.purge.wizard.menu').get_wizard_action()"
```

**Comment by AaronHForgeFlow** [comment]
> @AaronHForgeFlow Is #1396 also included in this PR?

not yet but I will. 

> I also get a traceback when opening some menu items:

I will check. Thanks!

**Comment by CasVissers-360ERP** [comment]
Some additional feedback:
1. 
image
Purge properties menu item name seems mixed up.
2. Purge obsolete menu items still gives an error in my own env:
  ```
File "/home/odoo/src/odoo/odoo/models.py", line 1747, in _add_missing_default_values
    defaults = self.default_get(list(missing_defaults))
  File "/home/odoo/src/user/community/database_cleanup/models/purge_wizard.py", line 45, in default_get
    res["purge_line_ids"] = self.find()
  File "/home/odoo/src/user/community/database_cleanup/models/purge_menus.py", line 50, in find
    menu.action.src_model and menu.action.src_model not in self.env
ValueError: <class 'AttributeError'>: "'ir.actions.act_window' object has no attribute 'src_model'" while evaluating
"action = env.get('cleanup.purge.wizard.menu').get_wizard_action()"
```
3. 
The model cleanup.create_indexes.line has no _description 
The model cleanup.purge.line.module has no _description 
The model cleanup.purge.line.data has no _description 
The model cleanup.purge.line.menu has no _description

**Comment by AaronHForgeFlow** [comment]
Thanks for the review @CasVissers  I will check as soon as I have some time for this, honestly, I have not tested it from the web, only from the orm. 

Regards.

**Comment by i-vyshnevska** [comment]
can you also add a description field to models, it is very login annoyng 

`
2020-05-24 16:52:10,613 298 WARNING odoodb odoo.models: The model cleanup.create_indexes.line has no _description 
2020-05-24 16:52:10,613 298 WARNING odoodb odoo.models: The model cleanup.purge.line.module has no _description 
2020-05-24 16:52:10,614 298 WARNING odoodb odoo.models: The model cleanup.purge.line.data has no _description 
2020-05-24 16:52:10,614 298 WARNING odoodb odoo.models: The model cleanup.purge.line.menu has no _description `

File: database_cleanup/models/purge_data.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+from ..identifier_adapter import IdentifierAdapter
+
+
+class CleanupPurgeLineData(models.TransientModel):
+    _inherit = "cleanup.purge.line"
+    _name = "cleanup.purge.line.data"
```
Comment (Tonow-c2c) [inline database_cleanup/models/purge_data.py:11]: please add one description

File: database_cleanup/models/purge_menus.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+# Copyright 2014-2016 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+# pylint: disable=consider-merging-classes-inherited
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class CleanupPurgeLineMenu(models.TransientModel):
+    _inherit = "cleanup.purge.line"
+    _name = "cleanup.purge.line.menu"
```
Comment (Tonow-c2c) [inline database_cleanup/models/purge_menus.py:10]: please add one description

File: database_cleanup/models/purge_modules.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                    this.unlink()
+                    continue
+            if this.model not in self.env:
+                this.unlink()
+        return super(IrModelData, self)._module_data_uninstall(modules_to_remove)
+
+
+class CleanupPurgeLineModule(models.TransientModel):
+    _inherit = "cleanup.purge.line"
+    _name = "cleanup.purge.line.module"
```
Comment (Tonow-c2c) [inline database_cleanup/models/purge_modules.py:35]: please add one description

**Comment by captivea-dcr** [comment]
@AaronHForgeFlow https://github.com/ForgeFlow/server-tools/pull/2 on your branch to (hopefully) fix travis

**Comment by AaronHForgeFlow** [comment]
Superseded by #1885

**Pull request closed without merging**